Scope 1 emissions are direct greenhouse (GHG) emissions that occur from sources that are controlled or owned by an organization (e.g., emissions associated with fuel combustion in boilers, furnaces, vehicles). Scope 2 emissions are indirect GHG emissions associated with the purchase of electricity, steam, heat, or cooling. 

take pres = 0.5 for calculation

In [1]:
import pandas as pd

df = pd.read_csv("2013_-_Global_500_Emissions_and_Response_Status.csv")

In [2]:
df

Company Name   Account Number        Country   Reporting Year  \
0                   VTB Bank           20356          Russia            2013   
1                        CEZ            3082  Czech Republic            2013   
2              Beiersdorf AG            1641         Germany            2013   
3           Roche Holding AG           15841     Switzerland            2013   
4            CSX Corporation            4120             USA            2013   
..                       ...             ...             ...             ...   
495  Yahoo Japan Corporation           20906           Japan            2013   
496             U.S. Bancorp           19593             USA            2013   
497                Rio Tinto           15829  United Kingdom            2013   
498            Hyundai Mobis            8713     South Korea            2013   
499      Cenovus Energy Inc.           29789          Canada            2013   

    Ticker Symbol          ISIN   Disclosure Score Performance Band  \
0          VTBR RU  RU000A0JP5V6               NaN              NaN   
1           CEZ CP  CZ0005112300               NaN              NaN   
2           BEI GR  DE0005200000              51.0                C   
3           ROG VX  CH0012032048              96.0                B   
4           CSX US  US1264081035              95.0                A   
..             ...           ...               ...              ...   
495        4689 JP  JP3933800009              15.0              NaN   
496         USB US  US9029733048              78.0                C   
497         RIO LN  GB0007188757              88.0                B   
498      012330 KS  KR7012330007              94.0                B   
499         CVE CN  CA15135U1093              94.0                B   

    Parent Account   Permission Response Status  Scope 1 (metric tonnes CO2e)  \
0               NaN         NaN              NR                           NaN   
1               NaN         NaN              NR                           NaN   
2               NaN      Public             AQ*                      22071.00   
3               NaN      Public             AQ*                     394063.00   
4               NaN      Public             AQ*                    5268905.00   
..              ...         ...             ...                           ...   
495             NaN  Not public             AQ*                           NaN   
496             NaN      Public             AQ*                      38055.00   
497             NaN      Public             AQ*                   26900000.00   
498             NaN      Public             AQ*                      40966.85   
499             NaN      Public             AQ*                    4657427.00   

     Scope 2 (metric tonnes CO2e)          Country Location  
0                             NaN    (61.52401, 105.318756)  
1                             NaN    (49.817492, 15.472962)  
2                        51571.00    (51.165691, 10.451526)  
3                       432103.00     (46.818188, 8.227512)  
4                       300170.00    (37.09024, -95.712891)  
..                            ...                       ...  
495                           NaN   (36.204824, 138.252924)  
496                     370672.00    (37.09024, -95.712891)  
497                   16400000.00    (55.378051, -3.435973)  
498                     291956.94   (35.907757, 127.766922)  
499                    1079646.00  (56.130366, -106.346771)  

[500 rows x 14 columns]

In [3]:
df2 = df[["Company Name ", "Ticker Symbol ","Scope 1 (metric tonnes CO2e)", "Scope 2 (metric tonnes CO2e)"]].copy()
df2

Company Name  Ticker Symbol   Scope 1 (metric tonnes CO2e)  \
0                   VTB Bank        VTBR RU                           NaN   
1                        CEZ         CEZ CP                           NaN   
2              Beiersdorf AG         BEI GR                      22071.00   
3           Roche Holding AG         ROG VX                     394063.00   
4            CSX Corporation         CSX US                    5268905.00   
..                       ...            ...                           ...   
495  Yahoo Japan Corporation        4689 JP                           NaN   
496             U.S. Bancorp         USB US                      38055.00   
497                Rio Tinto         RIO LN                   26900000.00   
498            Hyundai Mobis      012330 KS                      40966.85   
499      Cenovus Energy Inc.         CVE CN                    4657427.00   

     Scope 2 (metric tonnes CO2e)  
0                             NaN  
1                             NaN  
2                        51571.00  
3                       432103.00  
4                       300170.00  
..                            ...  
495                           NaN  
496                     370672.00  
497                   16400000.00  
498                     291956.94  
499                    1079646.00  

[500 rows x 4 columns]

In [4]:
df3=df2.dropna()
df3

Company Name  Ticker Symbol   Scope 1 (metric tonnes CO2e)  \
2          Beiersdorf AG         BEI GR                  2.207100e+04   
3       Roche Holding AG         ROG VX                  3.940630e+05   
4        CSX Corporation         CSX US                  5.268905e+06   
5               Centrica         CNA LN                  7.230344e+06   
7          Pernod Ricard          RI FP                  2.833450e+05   
..                   ...            ...                           ...   
494           Holcim Ltd        HOLN VX                  1.021021e+08   
496         U.S. Bancorp         USB US                  3.805500e+04   
497            Rio Tinto         RIO LN                  2.690000e+07   
498        Hyundai Mobis      012330 KS                  4.096685e+04   
499  Cenovus Energy Inc.         CVE CN                  4.657427e+06   

     Scope 2 (metric tonnes CO2e)  
2                        51571.00  
3                       432103.00  
4                       300170.00  
5                       166476.00  
7                        92502.00  
..                            ...  
494                    6600687.00  
496                     370672.00  
497                   16400000.00  
498                     291956.94  
499                    1079646.00  

[335 rows x 4 columns]

In [5]:
df3["scope 1 + scope 2"] = df3.apply(lambda x: x["Scope 1 (metric tonnes CO2e)"]+x["Scope 2 (metric tonnes CO2e)"], axis=1)
df3

ipykernel_launcher:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


Company Name  Ticker Symbol   Scope 1 (metric tonnes CO2e)  \
2          Beiersdorf AG         BEI GR                  2.207100e+04   
3       Roche Holding AG         ROG VX                  3.940630e+05   
4        CSX Corporation         CSX US                  5.268905e+06   
5               Centrica         CNA LN                  7.230344e+06   
7          Pernod Ricard          RI FP                  2.833450e+05   
..                   ...            ...                           ...   
494           Holcim Ltd        HOLN VX                  1.021021e+08   
496         U.S. Bancorp         USB US                  3.805500e+04   
497            Rio Tinto         RIO LN                  2.690000e+07   
498        Hyundai Mobis      012330 KS                  4.096685e+04   
499  Cenovus Energy Inc.         CVE CN                  4.657427e+06   

     Scope 2 (metric tonnes CO2e)  scope 1 + scope 2  
2                        51571.00       7.364200e+04  
3                       432103.00       8.261660e+05  
4                       300170.00       5.569075e+06  
5                       166476.00       7.396820e+06  
7                        92502.00       3.758470e+05  
..                            ...                ...  
494                    6600687.00       1.087028e+08  
496                     370672.00       4.087270e+05  
497                   16400000.00       4.330000e+07  
498                     291956.94       3.329238e+05  
499                    1079646.00       5.737073e+06  

[335 rows x 5 columns]

In [6]:
def normalize_z(df):
    dfout = df.copy()
    dfout = (df - df.mean(axis=0)) / df.std(axis=0)
    return dfout

In [7]:
df3["sustainability index"] = 100-((df3["scope 1 + scope 2"]-df3["scope 1 + scope 2"].min(axis=0))/(df3["scope 1 + scope 2"].max(axis=0)-df3["scope 1 + scope 2"].min(axis=0))*100)

ipykernel_launcher:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [8]:
df3

Company Name  Ticker Symbol   Scope 1 (metric tonnes CO2e)  \
2          Beiersdorf AG         BEI GR                  2.207100e+04   
3       Roche Holding AG         ROG VX                  3.940630e+05   
4        CSX Corporation         CSX US                  5.268905e+06   
5               Centrica         CNA LN                  7.230344e+06   
7          Pernod Ricard          RI FP                  2.833450e+05   
..                   ...            ...                           ...   
494           Holcim Ltd        HOLN VX                  1.021021e+08   
496         U.S. Bancorp         USB US                  3.805500e+04   
497            Rio Tinto         RIO LN                  2.690000e+07   
498        Hyundai Mobis      012330 KS                  4.096685e+04   
499  Cenovus Energy Inc.         CVE CN                  4.657427e+06   

     Scope 2 (metric tonnes CO2e)  scope 1 + scope 2  sustainability index  
2                        51571.00       7.364200e+04             99.960903  
3                       432103.00       8.261660e+05             99.551095  
4                       300170.00       5.569075e+06             96.968207  
5                       166476.00       7.396820e+06             95.972856  
7                        92502.00       3.758470e+05             99.796329  
..                            ...                ...                   ...  
494                    6600687.00       1.087028e+08             40.803760  
496                     370672.00       4.087270e+05             99.778423  
497                   16400000.00       4.330000e+07             76.420745  
498                     291956.94       3.329238e+05             99.819704  
499                    1079646.00       5.737073e+06             96.876719  

[335 rows x 6 columns]

In [9]:
def clean_symbol(symbol):
    l = symbol.split()
    return l[0]

df3["Symbol"] = df3.apply(lambda x: clean_symbol(x["Ticker Symbol "]), axis=1)
df3

ipykernel_launcher:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


Company Name  Ticker Symbol   Scope 1 (metric tonnes CO2e)  \
2          Beiersdorf AG         BEI GR                  2.207100e+04   
3       Roche Holding AG         ROG VX                  3.940630e+05   
4        CSX Corporation         CSX US                  5.268905e+06   
5               Centrica         CNA LN                  7.230344e+06   
7          Pernod Ricard          RI FP                  2.833450e+05   
..                   ...            ...                           ...   
494           Holcim Ltd        HOLN VX                  1.021021e+08   
496         U.S. Bancorp         USB US                  3.805500e+04   
497            Rio Tinto         RIO LN                  2.690000e+07   
498        Hyundai Mobis      012330 KS                  4.096685e+04   
499  Cenovus Energy Inc.         CVE CN                  4.657427e+06   

     Scope 2 (metric tonnes CO2e)  scope 1 + scope 2  sustainability index  \
2                        51571.00       7.364200e+04             99.960903   
3                       432103.00       8.261660e+05             99.551095   
4                       300170.00       5.569075e+06             96.968207   
5                       166476.00       7.396820e+06             95.972856   
7                        92502.00       3.758470e+05             99.796329   
..                            ...                ...                   ...   
494                    6600687.00       1.087028e+08             40.803760   
496                     370672.00       4.087270e+05             99.778423   
497                   16400000.00       4.330000e+07             76.420745   
498                     291956.94       3.329238e+05             99.819704   
499                    1079646.00       5.737073e+06             96.876719   

     Symbol  
2       BEI  
3       ROG  
4       CSX  
5       CNA  
7        RI  
..      ...  
494    HOLN  
496     USB  
497     RIO  
498  012330  
499     CVE  

[335 rows x 7 columns]

In [10]:
df_sus = df3[["Company Name ", "Symbol","sustainability index"]]
df_sus

Company Name   Symbol  sustainability index
2          Beiersdorf AG     BEI             99.960903
3       Roche Holding AG     ROG             99.551095
4        CSX Corporation     CSX             96.968207
5               Centrica     CNA             95.972856
7          Pernod Ricard      RI             99.796329
..                   ...     ...                   ...
494           Holcim Ltd    HOLN             40.803760
496         U.S. Bancorp     USB             99.778423
497            Rio Tinto     RIO             76.420745
498        Hyundai Mobis  012330             99.819704
499  Cenovus Energy Inc.     CVE             96.876719

[335 rows x 3 columns]

In [11]:
company = df_sus["Company Name "]

In [12]:
company

2            Beiersdorf AG
3         Roche Holding AG
4          CSX Corporation
5                 Centrica
7            Pernod Ricard
              ...         
494             Holcim Ltd
496           U.S. Bancorp
497              Rio Tinto
498          Hyundai Mobis
499    Cenovus Energy Inc.
Name: Company Name , Length: 335, dtype: object

In [13]:
l = list(company)

In [14]:
l

['Beiersdorf AG',
 'Roche Holding AG',
 'CSX Corporation',
 'Centrica',
 'Pernod Ricard',
 'Itaú Unibanco Holding S.A.',
 'Iberdrola SA',
 '3M Company',
 'Brookfield Asset Management Inc.',
 'Prudential Financial, Inc.',
 'Statoil ASA',
 'Enbridge Inc.',
 'Chunghwa Telecom',
 'National Australia Bank',
 'adidas AG',
 'Total',
 'CLP Holdings Limited',
 'Intesa Sanpaolo S.p.A',
 'Samsung Electronics',
 'The Home Depot, Inc.',
 'Las Vegas Sands Corporation',
 'GlaxoSmithKline',
 'Air Liquide',
 'Sasol Limited',
 'Bank of America',
 'Vinci',
 'Standard Chartered',
 'Atlas Copco',
 'Swisscom',
 'Mondelez International Inc',
 'Teva Pharmaceutical Industries Ltd',
 'Associated British Foods',
 'KDDI Corporation',
 'Shin-Etsu Chemical Co., Ltd.',
 'Banco Santander',
 'FedEx Corporation',
 'Royal Philips',
 'Union Pacific Corporation',
 'Franklin Resources, Inc.',
 'Ambev - Cia de Bebidas das Américas',
 'Taiwan Semiconductor Manufacturing',
 'AXA Group',
 'Canon Inc.',
 'Swiss Re',
 'DIRECTV',

In [15]:
stock_df = pd.read_csv("stock.csv")
stock_df

Symbol                                               Name Last Sale  \
0         A             Agilent Technologies Inc. Common Stock   $137.51   
1        AA                    Alcoa Corporation Common Stock     $56.21   
2       AAC  Ares Acquisition Corporation Class A Ordinary ...     $9.74   
3      AACG   ATA Creativity Global American Depositary Shares     $1.75   
4      AACI            Armada Acquisition Corp. I Common Stock     $9.81   
...     ...                                                ...       ...   
8399    ZWS      Zurn Water Solutions Corporation Common Stock    $31.51   
8400     ZY                         Zymergen Inc. Common Stock     $4.86   
8401   ZYME                       Zymeworks Inc. Common Shares    $11.21   
8402   ZYNE          Zynerba Pharmaceuticals Inc. Common Stock     $2.37   
8403   ZYXI                            Zynex Inc. Common Stock     $8.12   

      Net Change % Change    Market Cap        Country  IPO Year    Volume  \
0          -1.97  -1.412%  4.152813e+10  United States    1999.0   1878562   
1          -5.04  -8.229%  1.051707e+10            NaN    2016.0  11018919   
2           0.00    0.00%  1.217500e+09            NaN    2021.0   1164821   
3          -0.08  -4.372%  5.490961e+07          China       NaN     87925   
4          -0.03  -0.305%  2.031602e+08  United States    2021.0    204116   
...          ...      ...           ...            ...       ...       ...   
8399       -0.47   -1.47%  3.823723e+09  United States    2012.0    874690   
8400       -0.33  -6.358%  4.976679e+08  United States    2021.0   1030300   
8401       -0.34  -2.944%  5.218699e+08         Canada    2017.0    953748   
8402       -0.01   -0.42%  9.768556e+07  United States    2015.0    922085   
8403       -0.22  -2.638%  3.256028e+08  United States       NaN    222091   

                Sector                                    Industry  
0        Capital Goods                         Electrical Products  
1     Basic Industries                          Metal Fabrications  
2              Finance                           Business Services  
3        Miscellaneous              Service to the Health Industry  
4                  NaN                                         NaN  
...                ...                                         ...  
8399  Public Utilities                                Water Supply  
8400  Basic Industries                      Industrial Specialties  
8401               NaN                                         NaN  
8402       Health Care  Biotechnology: Pharmaceutical Preparations  
8403       Health Care                       Other Pharmaceuticals  

[8404 rows x 11 columns]

In [16]:
def clean_price(price):
    return price[1::]

stock_df["Price"] = stock_df.apply(lambda x: clean_price(x["Last Sale"]), axis=1)
stock_df

Symbol                                               Name Last Sale  \
0         A             Agilent Technologies Inc. Common Stock   $137.51   
1        AA                    Alcoa Corporation Common Stock     $56.21   
2       AAC  Ares Acquisition Corporation Class A Ordinary ...     $9.74   
3      AACG   ATA Creativity Global American Depositary Shares     $1.75   
4      AACI            Armada Acquisition Corp. I Common Stock     $9.81   
...     ...                                                ...       ...   
8399    ZWS      Zurn Water Solutions Corporation Common Stock    $31.51   
8400     ZY                         Zymergen Inc. Common Stock     $4.86   
8401   ZYME                       Zymeworks Inc. Common Shares    $11.21   
8402   ZYNE          Zynerba Pharmaceuticals Inc. Common Stock     $2.37   
8403   ZYXI                            Zynex Inc. Common Stock     $8.12   

      Net Change % Change    Market Cap        Country  IPO Year    Volume  \
0          -1.97  -1.412%  4.152813e+10  United States    1999.0   1878562   
1          -5.04  -8.229%  1.051707e+10            NaN    2016.0  11018919   
2           0.00    0.00%  1.217500e+09            NaN    2021.0   1164821   
3          -0.08  -4.372%  5.490961e+07          China       NaN     87925   
4          -0.03  -0.305%  2.031602e+08  United States    2021.0    204116   
...          ...      ...           ...            ...       ...       ...   
8399       -0.47   -1.47%  3.823723e+09  United States    2012.0    874690   
8400       -0.33  -6.358%  4.976679e+08  United States    2021.0   1030300   
8401       -0.34  -2.944%  5.218699e+08         Canada    2017.0    953748   
8402       -0.01   -0.42%  9.768556e+07  United States    2015.0    922085   
8403       -0.22  -2.638%  3.256028e+08  United States       NaN    222091   

                Sector                                    Industry   Price  
0        Capital Goods                         Electrical Products  137.51  
1     Basic Industries                          Metal Fabrications   56.21  
2              Finance                           Business Services    9.74  
3        Miscellaneous              Service to the Health Industry    1.75  
4                  NaN                                         NaN    9.81  
...                ...                                         ...     ...  
8399  Public Utilities                                Water Supply   31.51  
8400  Basic Industries                      Industrial Specialties    4.86  
8401               NaN                                         NaN   11.21  
8402       Health Care  Biotechnology: Pharmaceutical Preparations    2.37  
8403       Health Care                       Other Pharmaceuticals    8.12  

[8404 rows x 12 columns]

In [17]:
def clean_beta(change_percent):
    try:
        num = (change_percent)[:-1]
        return float(num) * 0.01
    except:
        return None

stock_df["Beta"] = stock_df.apply(lambda x: clean_beta(x["% Change"]), axis=1)


In [18]:
stock_df_info = stock_df[["Symbol", "Name", "Market Cap", "Price", "Beta"]]
stock_df_info

Symbol                                               Name    Market Cap  \
0         A             Agilent Technologies Inc. Common Stock  4.152813e+10   
1        AA                    Alcoa Corporation Common Stock   1.051707e+10   
2       AAC  Ares Acquisition Corporation Class A Ordinary ...  1.217500e+09   
3      AACG   ATA Creativity Global American Depositary Shares  5.490961e+07   
4      AACI            Armada Acquisition Corp. I Common Stock  2.031602e+08   
...     ...                                                ...           ...   
8399    ZWS      Zurn Water Solutions Corporation Common Stock  3.823723e+09   
8400     ZY                         Zymergen Inc. Common Stock  4.976679e+08   
8401   ZYME                       Zymeworks Inc. Common Shares  5.218699e+08   
8402   ZYNE          Zynerba Pharmaceuticals Inc. Common Stock  9.768556e+07   
8403   ZYXI                            Zynex Inc. Common Stock  3.256028e+08   

       Price     Beta  
0     137.51 -0.01412  
1      56.21 -0.08229  
2       9.74  0.00000  
3       1.75 -0.04372  
4       9.81 -0.00305  
...      ...      ...  
8399   31.51 -0.01470  
8400    4.86 -0.06358  
8401   11.21 -0.02944  
8402    2.37 -0.00420  
8403    8.12 -0.02638  

[8404 rows x 5 columns]

In [19]:
stock_df_info.dropna()

Symbol                                               Name    Market Cap  \
0         A             Agilent Technologies Inc. Common Stock  4.152813e+10   
1        AA                    Alcoa Corporation Common Stock   1.051707e+10   
2       AAC  Ares Acquisition Corporation Class A Ordinary ...  1.217500e+09   
3      AACG   ATA Creativity Global American Depositary Shares  5.490961e+07   
4      AACI            Armada Acquisition Corp. I Common Stock  2.031602e+08   
...     ...                                                ...           ...   
8399    ZWS      Zurn Water Solutions Corporation Common Stock  3.823723e+09   
8400     ZY                         Zymergen Inc. Common Stock  4.976679e+08   
8401   ZYME                       Zymeworks Inc. Common Shares  5.218699e+08   
8402   ZYNE          Zynerba Pharmaceuticals Inc. Common Stock  9.768556e+07   
8403   ZYXI                            Zynex Inc. Common Stock  3.256028e+08   

       Price     Beta  
0     137.51 -0.01412  
1      56.21 -0.08229  
2       9.74  0.00000  
3       1.75 -0.04372  
4       9.81 -0.00305  
...      ...      ...  
8399   31.51 -0.01470  
8400    4.86 -0.06358  
8401   11.21 -0.02944  
8402    2.37 -0.00420  
8403    8.12 -0.02638  

[7926 rows x 5 columns]

In [20]:
df_sus_stock=df_sus.merge(stock_df_info, how="inner", on="Symbol")

In [21]:
df_sus_stock

Company Name  Symbol  sustainability index  \
0               Roche Holding AG    ROG             99.551095   
1                CSX Corporation    CSX             96.968207   
2                       Centrica    CNA             95.972856   
3                     3M Company    MMM             96.314209   
4     Prudential Financial, Inc.    PRU             99.959781   
..                           ...    ...                   ...   
179  Philip Morris International     PM             99.524041   
180                   Ventas Inc    VTR             99.806179   
181                 U.S. Bancorp    USB             99.778423   
182                    Rio Tinto    RIO             76.420745   
183          Cenovus Energy Inc.    CVE             96.876719   

                                             Name    Market Cap   Price  \
0                 Rogers Corporation Common Stock  5.108822e+09  272.76   
1                    CSX Corporation Common Stock  7.563323e+10   34.10   
2          CNA Financial Corporation Common Stock  1.213521e+10   44.72   
3                         3M Company Common Stock  9.949005e+10  172.65   
4          Prudential Financial Inc. Common Stock  4.158378e+10  110.01   
..                                            ...           ...     ...   
179  Philip Morris International Inc Common Stock  1.602287e+11  102.92   
180                      Ventas Inc. Common Stock  2.094875e+10   52.48   
181                     U.S. Bancorp Common Stock  8.170215e+10   55.10   
182                    Rio Tinto Plc Common Stock  1.204149e+11   74.36   
183               Cenovus Energy Inc Common Stock  2.834837e+10   14.05   

        Beta  
0   -0.00095  
1   -0.03235  
2   -0.01433  
3   -0.00530  
4   -0.02152  
..       ...  
179  0.00882  
180  0.02220  
181 -0.03231  
182 -0.02363  
183 -0.04552  

[184 rows x 7 columns]

In [22]:
def cal_Ri(beta, r_market):
    return 0.035+float(beta)*(float(r_market)-0.035)

df_sus_stock["R_i"] = df_sus_stock.apply(lambda x: cal_Ri(x["Beta"], x["Price"]),axis=1)
df_sus_stock


Company Name  Symbol  sustainability index  \
0               Roche Holding AG    ROG             99.551095   
1                CSX Corporation    CSX             96.968207   
2                       Centrica    CNA             95.972856   
3                     3M Company    MMM             96.314209   
4     Prudential Financial, Inc.    PRU             99.959781   
..                           ...    ...                   ...   
179  Philip Morris International     PM             99.524041   
180                   Ventas Inc    VTR             99.806179   
181                 U.S. Bancorp    USB             99.778423   
182                    Rio Tinto    RIO             76.420745   
183          Cenovus Energy Inc.    CVE             96.876719   

                                             Name    Market Cap   Price  \
0                 Rogers Corporation Common Stock  5.108822e+09  272.76   
1                    CSX Corporation Common Stock  7.563323e+10   34.10   
2          CNA Financial Corporation Common Stock  1.213521e+10   44.72   
3                         3M Company Common Stock  9.949005e+10  172.65   
4          Prudential Financial Inc. Common Stock  4.158378e+10  110.01   
..                                            ...           ...     ...   
179  Philip Morris International Inc Common Stock  1.602287e+11  102.92   
180                      Ventas Inc. Common Stock  2.094875e+10   52.48   
181                     U.S. Bancorp Common Stock  8.170215e+10   55.10   
182                    Rio Tinto Plc Common Stock  1.204149e+11   74.36   
183               Cenovus Energy Inc Common Stock  2.834837e+10   14.05   

        Beta       R_i  
0   -0.00095 -0.224089  
1   -0.03235 -1.067003  
2   -0.01433 -0.605336  
3   -0.00530 -0.879860  
4   -0.02152 -2.331662  
..       ...       ...  
179  0.00882  0.942446  
180  0.02220  1.199279  
181 -0.03231 -1.744150  
182 -0.02363 -1.721300  
183 -0.04552 -0.602963  

[184 rows x 8 columns]

In [23]:
sigma_square = df_sus_stock["Price"].apply(lambda x: float(x)).var()
sigma_square

44640.97668193083

In [24]:
import numpy as np
pres = 0.5
def cal_signal(pres, sigma_square, R_i):
    return pres*np.log(sigma_square)+(2-pres)*R_i

df_sus_stock["signal"] = df_sus_stock.apply(lambda x: cal_signal(pres, sigma_square, x["R_i"]), axis=1)
df_sus_stock

Company Name  Symbol  sustainability index  \
0               Roche Holding AG    ROG             99.551095   
1                CSX Corporation    CSX             96.968207   
2                       Centrica    CNA             95.972856   
3                     3M Company    MMM             96.314209   
4     Prudential Financial, Inc.    PRU             99.959781   
..                           ...    ...                   ...   
179  Philip Morris International     PM             99.524041   
180                   Ventas Inc    VTR             99.806179   
181                 U.S. Bancorp    USB             99.778423   
182                    Rio Tinto    RIO             76.420745   
183          Cenovus Energy Inc.    CVE             96.876719   

                                             Name    Market Cap   Price  \
0                 Rogers Corporation Common Stock  5.108822e+09  272.76   
1                    CSX Corporation Common Stock  7.563323e+10   34.10   
2          CNA Financial Corporation Common Stock  1.213521e+10   44.72   
3                         3M Company Common Stock  9.949005e+10  172.65   
4          Prudential Financial Inc. Common Stock  4.158378e+10  110.01   
..                                            ...           ...     ...   
179  Philip Morris International Inc Common Stock  1.602287e+11  102.92   
180                      Ventas Inc. Common Stock  2.094875e+10   52.48   
181                     U.S. Bancorp Common Stock  8.170215e+10   55.10   
182                    Rio Tinto Plc Common Stock  1.204149e+11   74.36   
183               Cenovus Energy Inc Common Stock  2.834837e+10   14.05   

        Beta       R_i    signal  
0   -0.00095 -0.224089  5.017071  
1   -0.03235 -1.067003  3.752700  
2   -0.01433 -0.605336  4.445200  
3   -0.00530 -0.879860  4.033414  
4   -0.02152 -2.331662  1.855711  
..       ...       ...       ...  
179  0.00882  0.942446  6.766872  
180  0.02220  1.199279  7.152122  
181 -0.03231 -1.744150  2.736979  
182 -0.02363 -1.721300  2.771254  
183 -0.04552 -0.602963  4.448760  

[184 rows x 9 columns]

In [25]:
df_sus_stock["profitability index"] = ((df_sus_stock["signal"]-df_sus_stock["signal"].min(axis=0))/(df_sus_stock["signal"].max(axis=0)-df_sus_stock["signal"].min(axis=0)))*100

In [26]:
df_sus_stock

Company Name  Symbol  sustainability index  \
0               Roche Holding AG    ROG             99.551095   
1                CSX Corporation    CSX             96.968207   
2                       Centrica    CNA             95.972856   
3                     3M Company    MMM             96.314209   
4     Prudential Financial, Inc.    PRU             99.959781   
..                           ...    ...                   ...   
179  Philip Morris International     PM             99.524041   
180                   Ventas Inc    VTR             99.806179   
181                 U.S. Bancorp    USB             99.778423   
182                    Rio Tinto    RIO             76.420745   
183          Cenovus Energy Inc.    CVE             96.876719   

                                             Name    Market Cap   Price  \
0                 Rogers Corporation Common Stock  5.108822e+09  272.76   
1                    CSX Corporation Common Stock  7.563323e+10   34.10   
2          CNA Financial Corporation Common Stock  1.213521e+10   44.72   
3                         3M Company Common Stock  9.949005e+10  172.65   
4          Prudential Financial Inc. Common Stock  4.158378e+10  110.01   
..                                            ...           ...     ...   
179  Philip Morris International Inc Common Stock  1.602287e+11  102.92   
180                      Ventas Inc. Common Stock  2.094875e+10   52.48   
181                     U.S. Bancorp Common Stock  8.170215e+10   55.10   
182                    Rio Tinto Plc Common Stock  1.204149e+11   74.36   
183               Cenovus Energy Inc Common Stock  2.834837e+10   14.05   

        Beta       R_i    signal  profitability index  
0   -0.00095 -0.224089  5.017071            93.519611  
1   -0.03235 -1.067003  3.752700            92.330548  
2   -0.01433 -0.605336  4.445200            92.981802  
3   -0.00530 -0.879860  4.033414            92.594543  
4   -0.02152 -2.331662  1.855711            90.546548  
..       ...       ...       ...                  ...  
179  0.00882  0.942446  6.766872            95.165191  
180  0.02220  1.199279  7.152122            95.527494  
181 -0.03231 -1.744150  2.736979            91.375326  
182 -0.02363 -1.721300  2.771254            91.407560  
183 -0.04552 -0.602963  4.448760            92.985150  

[184 rows x 10 columns]

In [27]:
result_df = df_sus_stock[["Company Name ", "Symbol", "Price","Market Cap","sustainability index", "profitability index"]]

In [28]:
result_df

Company Name  Symbol   Price    Market Cap  \
0               Roche Holding AG    ROG  272.76  5.108822e+09   
1                CSX Corporation    CSX   34.10  7.563323e+10   
2                       Centrica    CNA   44.72  1.213521e+10   
3                     3M Company    MMM  172.65  9.949005e+10   
4     Prudential Financial, Inc.    PRU  110.01  4.158378e+10   
..                           ...    ...     ...           ...   
179  Philip Morris International     PM  102.92  1.602287e+11   
180                   Ventas Inc    VTR   52.48  2.094875e+10   
181                 U.S. Bancorp    USB   55.10  8.170215e+10   
182                    Rio Tinto    RIO   74.36  1.204149e+11   
183          Cenovus Energy Inc.    CVE   14.05  2.834837e+10   

     sustainability index  profitability index  
0               99.551095            93.519611  
1               96.968207            92.330548  
2               95.972856            92.981802  
3               96.314209            92.594543  
4               99.959781            90.546548  
..                    ...                  ...  
179             99.524041            95.165191  
180             99.806179            95.527494  
181             99.778423            91.375326  
182             76.420745            91.407560  
183             96.876719            92.985150  

[184 rows x 6 columns]

In [31]:
desired_s = 96
desired_p = 90

import math
def cal_similarity(company_s, company_p):
    return math.sqrt((company_s-desired_s)**2+(company_p-desired_p)**2)


result_df["absolute distance"] = result_df.apply(lambda x: cal_similarity(x["sustainability index"],x["profitability index"]), axis=1)
result_df

ipykernel_launcher:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


Company Name  Symbol   Price    Market Cap  \
0               Roche Holding AG    ROG  272.76  5.108822e+09   
1                CSX Corporation    CSX   34.10  7.563323e+10   
2                       Centrica    CNA   44.72  1.213521e+10   
3                     3M Company    MMM  172.65  9.949005e+10   
4     Prudential Financial, Inc.    PRU  110.01  4.158378e+10   
..                           ...    ...     ...           ...   
179  Philip Morris International     PM  102.92  1.602287e+11   
180                   Ventas Inc    VTR   52.48  2.094875e+10   
181                 U.S. Bancorp    USB   55.10  8.170215e+10   
182                    Rio Tinto    RIO   74.36  1.204149e+11   
183          Cenovus Energy Inc.    CVE   14.05  2.834837e+10   

     sustainability index  profitability index  absolute distance  
0               99.551095            93.519611           4.999793  
1               96.968207            92.330548           2.523664  
2               95.972856            92.981802           2.981925  
3               96.314209            92.594543           2.613500  
4               99.959781            90.546548           3.997322  
..                    ...                  ...                ...  
179             99.524041            95.165191           6.252844  
180             99.806179            95.527494           6.711199  
181             99.778423            91.375326           4.020946  
182             76.420745            91.407560          19.629785  
183             96.876719            92.985150           3.111230  

[184 rows x 7 columns]

In [32]:
# now what i am doing is to make the absolute distance to be in between 0 to 100 scale
min_dis = result_df["absolute distance"].min(axis=0)
max_dis = result_df["absolute distance"].max(axis=0)
result_df["result similarity"] = 100 - ((result_df["absolute distance"]-min_dis)/(max_dis-min_dis))*100
result_df

ipykernel_launcher:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


Company Name  Symbol   Price    Market Cap  \
0               Roche Holding AG    ROG  272.76  5.108822e+09   
1                CSX Corporation    CSX   34.10  7.563323e+10   
2                       Centrica    CNA   44.72  1.213521e+10   
3                     3M Company    MMM  172.65  9.949005e+10   
4     Prudential Financial, Inc.    PRU  110.01  4.158378e+10   
..                           ...    ...     ...           ...   
179  Philip Morris International     PM  102.92  1.602287e+11   
180                   Ventas Inc    VTR   52.48  2.094875e+10   
181                 U.S. Bancorp    USB   55.10  8.170215e+10   
182                    Rio Tinto    RIO   74.36  1.204149e+11   
183          Cenovus Energy Inc.    CVE   14.05  2.834837e+10   

     sustainability index  profitability index  absolute distance  \
0               99.551095            93.519611           4.999793   
1               96.968207            92.330548           2.523664   
2               95.972856            92.981802           2.981925   
3               96.314209            92.594543           2.613500   
4               99.959781            90.546548           3.997322   
..                    ...                  ...                ...   
179             99.524041            95.165191           6.252844   
180             99.806179            95.527494           6.711199   
181             99.778423            91.375326           4.020946   
182             76.420745            91.407560          19.629785   
183             96.876719            92.985150           3.111230   

     result similarity  
0            95.478077  
1            98.209682  
2            97.704139  
3            98.110577  
4            96.583979  
..                 ...  
179          94.095742  
180          93.590097  
181          96.557918  
182          79.338630  
183          97.561493  

[184 rows x 8 columns]

In [29]:
result_df.to_csv('result.csv', index=False)

take this example: 

- Roche Holding AG	
- company_s: 99.551095	
- company_p: 93.519611
- user input: desired_s=98, desired_p=